content/style/go.md

```diff
@@ -1,3 +1,1659 @@
+    result[k] = v
+  }
+  return result
+}
+
+// Snapshot is now a copy.
+snapshot := stats.Snapshot()
+```
+
+</td></tr>
+</tbody></table>
+
+### Start Enums at One
+
+The standard way of introducing enumerations in Go is to declare a custom type and a `const` group with `iota`.
+Since variables have a 0 default value, you should usually start your enums on a non-zero value.
+
+<table>
+<thead><tr><th>Bad</th><th>Good</th></tr></thead>
+<tbody>
+<tr><td>
+
+```go
+type Operation int
+
+const (
+  Add Operation = iota
+  Subtract
+  Multiply
+)
+
+// Add=0, Subtract=1, Multiply=2
+```
+
+</td><td>
+
+```go
+type Operation int
+
+const (
+  Add Operation = iota + 1
+  Subtract
+  Multiply
+)
+
+// Add=1, Subtract=2, Multiply=3
+```
+
+</td></tr>
+</tbody></table>
+
+There are cases where using the zero value makes sense, for example when the zero value case is the desirable default behavior.
+
+```go
+type LogOutput int
+
+const (
+  LogToStdout LogOutput = iota
+  LogToFile
+  LogToRemote
+)
+
+// LogToStdout=0, LogToFile=1, LogToRemote=2
+```
+
+### Error Types
+
+There are various options for declaring errors:
+
+* [`errors.New`] for errors with simple static strings
+* [`fmt.Errorf`] for formatted error strings
+* Custom types that implement an `Error()` method
+* Wrapped errors using [`"pkg/errors".Wrap`]
+
+When returning errors, consider the following to determine the best choice:
+
+* Is this a simple error that needs no extra information? If so, [`errors.New`] should suffice.
+* Do the clients need to detect and handle this error? If so, you should use a custom type, and implement the `Error()` method.
+* Are you propagating an error returned by a downstream function? If so, check the [section on error wrapping](#error-wrapping).
+* Otherwise, [`fmt.Errorf`] is okay.
+
+  [`errors.New`]: https://golang.org/pkg/errors/#New
+  [`fmt.Errorf`]: https://golang.org/pkg/fmt/#Errorf
+  [`"pkg/errors".Wrap`]: https://godoc.org/github.com/pkg/errors#Wrap
+
+If the client needs to detect the error, and you have created a simple error using [`errors.New`], use a var for the error.
+
+<table>
+<thead><tr><th>Bad</th><th>Good</th></tr></thead>
+<tbody>
+<tr><td>
+
+```go
+// package foo
+
+func Open() error {
+  return errors.New("could not open")
+}
+
+// package bar
+
+func use() {
+  if err := foo.Open(); err != nil {
+    if err.Error() == "could not open" {
+      // handle
+    } else {
+      panic("unknown error")
+    }
+  }
+}
+```
+
+</td><td>
+
+```go
+// package foo
+
+var ErrCouldNotOpen = errors.New("could not open")
+
+func Open() error {
+  return ErrCouldNotOpen
+}
+
+// package bar
+
+if err := foo.Open(); err != nil {
+  if errors.Is(err, foo.ErrCouldNotOpen) {
+    // handle
+  } else {
+    panic("unknown error")
+  }
+}
+```
+
+</td></tr>
+</tbody></table>
+
+If you have an error that clients may need to detect, and you would like to add more information to it (e.g., it is not a static string), then you should use a custom type.
```

Comment:
I think this can be worded more precisely. There are cases where the regular `fmt.Errorf` is okay, and we do use it semi-regularly.

For example - `fmt.Errorf("invalid height (given: %d, first: %d, last: %d)", height, first, last)`, or  `fmt.Errorf("could not retrieve transaction (%x): %w", txID, err)`.

It's more that, when the client may need to have *structured* information about the specific error, we should have a custom type

Reply:
I agree, in our case most often `fmt.Errorf` is okay, since it's the _default_ case if nothing else makes sense, so as the sentence states, using custom error types makes sense when it should be detected by the caller. This means the caller uses `errors.Is` to recognize it, and can access the attributes if needed.

For something where no extra metadata is needed, using a sentinel error is fine. WDYT?

Reply:
Yeah, you're right, the sentence does say when it's an error that the client needs to detect AND you'd like to add more information to it :)